xds/internal/balancer/clusterresolver/config.go

```diff
@@ -18,21 +18,137 @@
 package clusterresolver
 
 import (
+	"bytes"
 	"encoding/json"
+	"fmt"
 
 	internalserviceconfig "google.golang.org/grpc/internal/serviceconfig"
 	"google.golang.org/grpc/serviceconfig"
-	"google.golang.org/grpc/xds/internal/balancer/clusterresolver/balancerconfig"
 )
 
+// DiscoveryMechanismType is the type of discovery mechanism.
+type DiscoveryMechanismType int
+
+const (
+	// DiscoveryMechanismTypeEDS is eds.
+	DiscoveryMechanismTypeEDS DiscoveryMechanismType = iota // `json:"EDS"`
+	// DiscoveryMechanismTypeLogicalDNS is DNS.
+	DiscoveryMechanismTypeLogicalDNS // `json:"LOGICAL_DNS"`
+)
+
+// MarshalJSON marshals a DiscoveryMechanismType to a quoted json string.
+//
+// This is necessary to handle enum (as strings) from JSON.
+//
+// Note that this needs to be defined on the type not pointer, otherwise the
+// variables of this type will marshal to int not string.
+func (t DiscoveryMechanismType) MarshalJSON() ([]byte, error) {
+	buffer := bytes.NewBufferString(`"`)
+	switch t {
+	case DiscoveryMechanismTypeEDS:
+		buffer.WriteString("EDS")
+	case DiscoveryMechanismTypeLogicalDNS:
+		buffer.WriteString("LOGICAL_DNS")
+	}
+	buffer.WriteString(`"`)
+	return buffer.Bytes(), nil
+}
+
+// UnmarshalJSON unmarshals a quoted json string to the DiscoveryMechanismType.
+func (t *DiscoveryMechanismType) UnmarshalJSON(b []byte) error {
+	var s string
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+	switch s {
+	case "EDS":
```

Comment:
Is this supposed to be case-sensitive?

Reply:
I think so?

This is an enum, not a policy name.
And we never manually generate JSON. This just needs to be consistent with the `MarshalJSON()`